test/schema.union.validation.test.js

```diff
@@ -0,0 +1,314 @@
+'use strict';
+
+const start = require('./common');
+const util = require('./util');
+
+const assert = require('assert');
+
+const mongoose = start.mongoose;
+
+describe('Union validation', function() {
+  let db;
+
+  before(async function() {
+    db = await start().asPromise();
+  });
+
+  after(async function() {
+    await db.close();
+  });
+
+  afterEach(() => db.deleteModel(/Test/));
+  afterEach(() => util.clearTestData(db));
+  afterEach(() => util.stopRemainingOps(db));
+
+  it('should validate required fields in union schemas', async function() {
```

Comment:
Given the complexity of objects in union types, I think we might just recommend using [embedded discriminators](https://thecodebarbarian.com/mongoose-4.8-embedded-discriminators.html) instead.